// vue-country-flag/dist/country-flag.ssr.js
'use strict';function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _slicedToArray(arr, i) {
  return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest();
}

function _arrayWithHoles(arr) {
  if (Array.isArray(arr)) return arr;
}

function _iterableToArrayLimit(arr, i) {
  if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return;
  var _arr = [];
  var _n = true;
  var _d = false;
  var _e = undefined;

  try {
    for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) {
      _arr.push(_s.value);

      if (i && _arr.length === i) break;
    }
  } catch (err) {
    _d = true;
    _e = err;
  } finally {
    try {
      if (!_n && _i["return"] != null) _i["return"]();
    } finally {
      if (_d) throw _e;
    }
  }

  return _arr;
}

function _unsupportedIterableToArray(o, minLen) {
  if (!o) return;
  if (typeof o === "string") return _arrayLikeToArray(o, minLen);
  var n = Object.prototype.toString.call(o).slice(8, -1);
  if (n === "Object" && o.constructor) n = o.constructor.name;
  if (n === "Map" || n === "Set") return Array.from(o);
  if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen);
}

function _arrayLikeToArray(arr, len) {
  if (len == null || len > arr.length) len = arr.length;

  for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i];

  return arr2;
}

function _nonIterableRest() {
  throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method.");
}//
//
//
//
var script = {
  name: 'CountryFlag',
  props: {
    country: {
      type: String,
      required: true
    },
    rounded: {
      type: Boolean,
      default: false
    },
    size: {
      type: String,
      validator: function validator(value) {
        return value === 'small' || value === 'normal' || value === 'big';
      },
      default: 'normal'
    },
    background: {
      type: String,
      default: ''
    }
  },
  computed: {
    //vue-country-flag is dynamically replaced through rollup-replace
    backgroundUrl: function backgroundUrl() {
      if (this.background && this.background != '') {
        return this.background;
      } else {
        {
          return "vue-country-flag/dist/flags.png";
        }
      }
    },
    flagIconClass: function flagIconClass() {
      var _ref;

      return _ref = {}, _defineProperty(_ref, "flag ".concat(this.flagIconCountry), true), _defineProperty(_ref, "rounded", this.rounded), _defineProperty(_ref, this.flagMargin, true), _ref;
    },
    flagIconCountry: function flagIconCountry() {
      var country = this.country.toLowerCase();
      country = country === 'ad' ? 'andorra' : country;
      return "f-".concat(country);
    },
    flagMargin: function flagMargin() {
      switch (this.size) {
        case 'small':
          return 'small-flag';

        case 'normal':
          return 'normal-flag';

        case 'big':
          return 'big-flag';

        default:
          return 'normal-flag';
      }
    }
  }
};function normalizeComponent(template, style, script, scopeId, isFunctionalTemplate, moduleIdentifier /* server only */, shadowMode, createInjector, createInjectorSSR, createInjectorShadow) {
    if (typeof shadowMode !== 'boolean') {
        createInjectorSSR = createInjector;
        createInjector = shadowMode;
        shadowMode = false;
    }
    // Vue.extend constructor export interop.
    const options = typeof script === 'function' ? script.options : script;
    // render functions
    if (template && template.render) {
        options.render = template.render;
        options.staticRenderFns = template.staticRenderFns;
        options._compiled = true;
        // functional template
        if (isFunctionalTemplate) {
            options.functional = true;
        }
    }
    // scopedId
    if (scopeId) {
        options._scopeId = scopeId;
    }
    let hook;
    if (moduleIdentifier) {
        // server build
        hook = function (context) {
            // 2.3 injection
            context =
                context || // cached call
                    (this.$vnode && this.$vnode.ssrContext) || // stateful
                    (this.parent && this.parent.$vnode && this.parent.$vnode.ssrContext); // functional
            // 2.2 with runInNewContext: true
            if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
                context = __VUE_SSR_CONTEXT__;
            }
            // inject component styles
            if (style) {
                style.call(this, createInjectorSSR(context));
            }
            // register component module identifier for async chunk inference
            if (context && context._registeredComponents) {
                context._registeredComponents.add(moduleIdentifier);
            }
        };
        // used by ssr in case component is cached and beforeCreate
        // never gets called
        options._ssrRegister = hook;
    }
    else if (style) {
        hook = shadowMode
            ? function (context) {
                style.call(this, createInjectorShadow(context, this.$root.$options.shadowRoot));
            }
            : function (context) {
                style.call(this, createInjector(context));
            };
    }
    if (hook) {
        if (options.functional) {
            // register for functional component in vue file
            const originalRender = options.render;
            options.render = function renderWithStyleInjection(h, context) {
                hook.call(context);
                return originalRender(h, context);
            };
        }
        else {
            // inject component registration as beforeCreate hook
            const existing = options.beforeCreate;
            options.beforeCreate = existing ? [].concat(existing, hook) : [hook];
        }
    }
    return script;
}function createInjectorSSR(context) {
    if (!context && typeof __VUE_SSR_CONTEXT__ !== 'undefined') {
        context = __VUE_SSR_CONTEXT__;
    }
    if (!context)
        return () => { };
    if (!('styles' in context)) {
        context._styles = context._styles || {};
        Object.defineProperty(context, 'styles', {
            enumerable: true,
            get: () => context._renderStyles(context._styles)
        });
        context._renderStyles = context._renderStyles || renderStyles;
    }
    return (id, style) => addStyle(id, style, context);
}
function addStyle(id, css, context) {
    const group =  css.media || 'default' ;
    const style = context._styles[group] || (context._styles[group] = { ids: [], css: '' });
    if (!style.ids.includes(id)) {
        style.media = css.media;
        style.ids.push(id);
        let code = css.source;
        style.css += code + '\n';
    }
}
function renderStyles(styles) {
    let css = '';
    for (const key in styles) {
        const style = styles[key];
        css +=
            '<style data-vue-ssr-id="' +
                Array.from(style.ids).join(' ') +
                '"' +
                (style.media ? ' media="' + style.media + '"' : '') +
                '>' +
                style.css +
                '</style>';
    }
    return css;
}/* script */
var __vue_script__ = script;
/* template */

var __vue_render__ = function __vue_render__() {
  var _vm = this;

  var _h = _vm.$createElement;

  var _c = _vm._self._c || _h;

  return _vm.country ? _c('span', {
    class: _vm.flagIconClass,
    style: {
      backgroundImage: "url('" + _vm.backgroundUrl + "')"
    }
  }, []) : _vm._e();
};

var __vue_staticRenderFns__ = [];
/* style */

var __vue_inject_styles__ = function __vue_inject_styles__(inject) {
  if (!inject) return;
  inject("data-v-569f794f_0", {
    source: ".small-flag[data-v-569f794f]{margin:-15px -22.5px;transform:scale(.25);-ms-transform:scale(.25);-webkit-transform:scale(.25);-moz-transform:scale(.25)}.normal-flag[data-v-569f794f]{margin:-10px -15px;transform:scale(.5);-ms-transform:scale(.5);-webkit-transform:scale(.5);-moz-transform:scale(.5)}.big-flag[data-v-569f794f]{margin:0}.small-flag.rounded[data-v-569f794f]{-moz-border-radius:15px;border-radius:15px}.normal-flag.rounded[data-v-569f794f]{-moz-border-radius:10px;border-radius:10px}.big-flag.rounded[data-v-569f794f]{-moz-border-radius:8px;border-radius:8px}.flag[data-v-569f794f]{width:60px;height:40px;display:inline-block}.f-and[data-v-569f794f],.f-andorra[data-v-569f794f]{background-position:0 -41px}.f-ae[data-v-569f794f],.f-are[data-v-569f794f]{background-position:0 -82px}.f-af[data-v-569f794f],.f-afg[data-v-569f794f]{background-position:0 -123px}.f-ag[data-v-569f794f],.f-atg[data-v-569f794f]{background-position:0 -164px}.f-ai[data-v-569f794f],.f-aia[data-v-569f794f]{background-position:0 -205px}.f-al[data-v-569f794f],.f-alb[data-v-569f794f]{background-position:0 -246px}.f-am[data-v-569f794f],.f-arm[data-v-569f794f]{background-position:0 -287px}.f-an[data-v-569f794f],.f-ant[data-v-569f794f]{background-position:0 -328px}.f-ago[data-v-569f794f],.f-ao[data-v-569f794f]{background-position:0 -369px}.f-aq[data-v-569f794f],.f-ata[data-v-569f794f]{background-position:0 -410px}.f-ar[data-v-569f794f],.f-arg[data-v-569f794f]{background-position:0 -451px}.f-as[data-v-569f794f],.f-asm[data-v-569f794f]{background-position:0 -492px}.f-at[data-v-569f794f],.f-aut[data-v-569f794f]{background-position:0 -533px}.f-au[data-v-569f794f],.f-aus[data-v-569f794f]{background-position:0 -574px}.f-abw[data-v-569f794f],.f-aw[data-v-569f794f]{background-position:0 -615px}.f-ala[data-v-569f794f],.f-ax[data-v-569f794f]{background-position:0 -656px}.f-az[data-v-569f794f],.f-aze[data-v-569f794f]{background-position:0 -697px}.f-ba[data-v-569f794f],.f-bih[data-v-569f794f]{background-position:0 -738px}.f-bb[data-v-569f794f],.f-brb[data-v-569f794f]{background-position:0 -779px}.f-bd[data-v-569f794f],.f-bgd[data-v-569f794f]{background-position:0 -820px}.f-be[data-v-569f794f],.f-bel[data-v-569f794f]{background-position:0 -861px}.f-bf[data-v-569f794f],.f-bfa[data-v-569f794f]{background-position:0 -902px}.f-bg[data-v-569f794f],.f-bgr[data-v-569f794f]{background-position:0 -943px}.f-bh[data-v-569f794f],.f-bhr[data-v-569f794f]{background-position:0 -984px}.f-bdi[data-v-569f794f],.f-bi[data-v-569f794f]{background-position:0 -1025px}.f-ben[data-v-569f794f],.f-bj[data-v-569f794f]{background-position:0 -1066px}.f-bl[data-v-569f794f],.f-blm[data-v-569f794f]{background-position:0 -1107px}.f-bm[data-v-569f794f],.f-bmu[data-v-569f794f]{background-position:0 -1148px}.f-bn[data-v-569f794f],.f-brn[data-v-569f794f]{background-position:0 -1189px}.f-bo[data-v-569f794f],.f-bol[data-v-569f794f]{background-position:0 -1230px}.f-bes[data-v-569f794f],.f-bq[data-v-569f794f]{background-position:0 -1271px}.f-br[data-v-569f794f],.f-bra[data-v-569f794f]{background-position:0 -1312px}.f-bhs[data-v-569f794f],.f-bs[data-v-569f794f]{background-position:0 -1353px}.f-bt[data-v-569f794f],.f-btn[data-v-569f794f]{background-position:0 -1394px}.f-bv[data-v-569f794f],.f-bvt[data-v-569f794f]{background-position:0 -1435px}.f-bw[data-v-569f794f],.f-bwa[data-v-569f794f]{background-position:0 -1476px}.f-blr[data-v-569f794f],.f-by[data-v-569f794f]{background-position:0 -1517px}.f-blz[data-v-569f794f],.f-bz[data-v-569f794f]{background-position:0 -1558px}.f-ca[data-v-569f794f],.f-can[data-v-569f794f]{background-position:0 -1599px}.f-cc[data-v-569f794f],.f-cck[data-v-569f794f]{background-position:0 -1640px}.f-cd[data-v-569f794f],.f-cod[data-v-569f794f]{background-position:0 -1681px}.f-caf[data-v-569f794f],.f-cf[data-v-569f794f]{background-position:0 -1722px}.f-cg[data-v-569f794f],.f-cog[data-v-569f794f]{background-position:0 -1763px}.f-ch[data-v-569f794f],.f-che[data-v-569f794f]{background-position:0 -1804px}.f-ci[data-v-569f794f],.f-civ[data-v-569f794f]{background-position:0 -1845px}.f-ck[data-v-569f794f],.f-cok[data-v-569f794f]{background-position:0 -1886px}.f-chl[data-v-569f794f],.f-cl[data-v-569f794f]{background-position:0 -1927px}.f-cm[data-v-569f794f],.f-cmr[data-v-569f794f]{background-position:0 -1968px}.f-chn[data-v-569f794f],.f-cn[data-v-569f794f]{background-position:0 -2009px}.f-co[data-v-569f794f],.f-col[data-v-569f794f]{background-position:0 -2050px}.f-cr[data-v-569f794f],.f-cri[data-v-569f794f]{background-position:0 -2091px}.f-cu[data-v-569f794f],.f-cub[data-v-569f794f]{background-position:0 -2132px}.f-cpv[data-v-569f794f],.f-cv[data-v-569f794f]{background-position:0 -2173px}.f-cuw[data-v-569f794f],.f-cw[data-v-569f794f]{background-position:0 -2214px}.f-cx[data-v-569f794f],.f-cxr[data-v-569f794f]{background-position:0 -2255px}.f-cy[data-v-569f794f],.f-cyp[data-v-569f794f]{background-position:0 -2296px}.f-cz[data-v-569f794f],.f-cze[data-v-569f794f]{background-position:0 -2337px}.f-de[data-v-569f794f],.f-deu[data-v-569f794f]{background-position:0 -2378px}.f-dj[data-v-569f794f],.f-dji[data-v-569f794f]{background-position:0 -2419px}.f-dk[data-v-569f794f],.f-dnk[data-v-569f794f]{background-position:0 -2460px}.f-dm[data-v-569f794f],.f-dma[data-v-569f794f]{background-position:0 -2501px}.f-do[data-v-569f794f],.f-dom[data-v-569f794f]{background-position:0 -2542px}.f-dz[data-v-569f794f],.f-dza[data-v-569f794f]{background-position:0 -2583px}.f-ec[data-v-569f794f],.f-ecu[data-v-569f794f]{background-position:0 -2624px}.f-ee[data-v-569f794f],.f-est[data-v-569f794f]{background-position:0 -2665px}.f-eg[data-v-569f794f],.f-egy[data-v-569f794f]{background-position:0 -2706px}.f-eh[data-v-569f794f],.f-esh[data-v-569f794f]{background-position:0 -2747px}.f-er[data-v-569f794f],.f-eri[data-v-569f794f]{background-position:0 -2788px}.f-es-ca[data-v-569f794f]{background-position:0 -2829px}.f-es[data-v-569f794f],.f-esp[data-v-569f794f]{background-position:0 -2870px}.f-et[data-v-569f794f],.f-eth[data-v-569f794f]{background-position:0 -2911px}.f-eu[data-v-569f794f]{background-position:0 -2952px}.f-fi[data-v-569f794f],.f-fin[data-v-569f794f]{background-position:0 -2993px}.f-fj[data-v-569f794f],.f-fji[data-v-569f794f]{background-position:0 -3034px}.f-fk[data-v-569f794f],.f-flk[data-v-569f794f]{background-position:0 -3075px}.f-fm[data-v-569f794f],.f-fsm[data-v-569f794f]{background-position:0 -3116px}.f-fo[data-v-569f794f],.f-fro[data-v-569f794f]{background-position:0 -3157px}.f-fr[data-v-569f794f],.f-fra[data-v-569f794f]{background-position:0 -3198px}.f-ga[data-v-569f794f],.f-gab[data-v-569f794f]{background-position:0 -3239px}.f-gb-eng[data-v-569f794f]{background-position:0 -3280px}.f-gb-nir[data-v-569f794f]{background-position:0 -3321px}.f-gb-sct[data-v-569f794f]{background-position:0 -3362px}.f-gb-wls[data-v-569f794f]{background-position:0 -3403px}.f-gb[data-v-569f794f],.f-gbr[data-v-569f794f]{background-position:0 -3444px}.f-gd[data-v-569f794f],.f-grd[data-v-569f794f]{background-position:0 -3485px}.f-ge[data-v-569f794f],.f-geo[data-v-569f794f]{background-position:0 -3526px}.f-gf[data-v-569f794f],.f-guf[data-v-569f794f]{background-position:0 -3567px}.f-gg[data-v-569f794f],.f-ggy[data-v-569f794f]{background-position:0 -3608px}.f-gh[data-v-569f794f],.f-gha[data-v-569f794f]{background-position:0 -3649px}.f-gi[data-v-569f794f],.f-gib[data-v-569f794f]{background-position:0 -3690px}.f-gl[data-v-569f794f],.f-grl[data-v-569f794f]{background-position:0 -3731px}.f-gm[data-v-569f794f],.f-gmb[data-v-569f794f]{background-position:0 -3772px}.f-gin[data-v-569f794f],.f-gn[data-v-569f794f]{background-position:0 -3813px}.f-glp[data-v-569f794f],.f-gp[data-v-569f794f]{background-position:0 -3854px}.f-gnq[data-v-569f794f],.f-gq[data-v-569f794f]{background-position:0 -3895px}.f-gr[data-v-569f794f],.f-grc[data-v-569f794f]{background-position:0 -3936px}.f-gs[data-v-569f794f],.f-sgs[data-v-569f794f]{background-position:0 -3977px}.f-gt[data-v-569f794f],.f-gtm[data-v-569f794f]{background-position:0 -4018px}.f-gu[data-v-569f794f],.f-gum[data-v-569f794f]{background-position:0 -4059px}.f-gnb[data-v-569f794f],.f-gw[data-v-569f794f]{background-position:0 -4100px}.f-guy[data-v-569f794f],.f-gy[data-v-569f794f]{background-position:0 -4141px}.f-hk[data-v-569f794f],.f-hkg[data-v-569f794f]{background-position:0 -4182px}.f-hm[data-v-569f794f],.f-hmd[data-v-569f794f]{background-position:0 -4223px}.f-hn[data-v-569f794f],.f-hnd[data-v-569f794f]{background-position:0 -4264px}.f-hr[data-v-569f794f],.f-hrv[data-v-569f794f]{background-position:0 -4305px}.f-ht[data-v-569f794f],.f-hti[data-v-569f794f]{background-position:0 -4346px}.f-hu[data-v-569f794f],.f-hun[data-v-569f794f]{background-position:0 -4387px}.f-id[data-v-569f794f],.f-idn[data-v-569f794f]{background-position:0 -4428px}.f-ie[data-v-569f794f],.f-irl[data-v-569f794f]{background-position:0 -4469px}.f-il[data-v-569f794f],.f-isr[data-v-569f794f]{background-position:0 -4510px}.f-im[data-v-569f794f],.f-imn[data-v-569f794f]{background-position:0 -4551px}.f-in[data-v-569f794f],.f-ind[data-v-569f794f]{background-position:0 -4592px}.f-io[data-v-569f794f],.f-iot[data-v-569f794f]{background-position:0 -4633px}.f-iq[data-v-569f794f],.f-irq[data-v-569f794f]{background-position:0 -4674px}.f-ir[data-v-569f794f],.f-irn[data-v-569f794f]{background-position:0 -4715px}.f-is[data-v-569f794f],.f-isl[data-v-569f794f]{background-position:0 -4756px}.f-it[data-v-569f794f],.f-ita[data-v-569f794f]{background-position:0 -4797px}.f-je[data-v-569f794f],.f-jey[data-v-569f794f]{background-position:0 -4838px}.f-jam[data-v-569f794f],.f-jm[data-v-569f794f]{background-position:0 -4879px}.f-jo[data-v-569f794f],.f-jor[data-v-569f794f]{background-position:0 -4920px}.f-jp[data-v-569f794f],.f-jpn[data-v-569f794f]{background-position:0 -4961px}.f-ke[data-v-569f794f],.f-ken[data-v-569f794f]{background-position:0 -5002px}.f-kg[data-v-569f794f],.f-kgz[data-v-569f794f]{background-position:0 -5043px}.f-kh[data-v-569f794f],.f-khm[data-v-569f794f]{background-position:0 -5084px}.f-ki[data-v-569f794f],.f-kir[data-v-569f794f]{background-position:0 -5125px}.f-com[data-v-569f794f],.f-km[data-v-569f794f]{background-position:0 -5166px}.f-kn[data-v-569f794f],.f-kna[data-v-569f794f]{background-position:0 -5207px}.f-kp[data-v-569f794f],.f-prk[data-v-569f794f]{background-position:0 -5248px}.f-kor[data-v-569f794f],.f-kr[data-v-569f794f]{background-position:0 -5289px}.f-kw[data-v-569f794f],.f-kwt[data-v-569f794f]{background-position:0 -5330px}.f-cym[data-v-569f794f],.f-ky[data-v-569f794f]{background-position:0 -5371px}.f-kaz[data-v-569f794f],.f-kz[data-v-569f794f]{background-position:0 -5412px}.f-la[data-v-569f794f],.f-lao[data-v-569f794f]{background-position:0 -5453px}.f-lb[data-v-569f794f],.f-lbn[data-v-569f794f]{background-position:0 -5494px}.f-lc[data-v-569f794f],.f-lca[data-v-569f794f]{background-position:0 -5535px}.f-li[data-v-569f794f],.f-lie[data-v-569f794f]{background-position:0 -5576px}.f-lk[data-v-569f794f],.f-lka[data-v-569f794f]{background-position:0 -5617px}.f-lbr[data-v-569f794f],.f-lr[data-v-569f794f]{background-position:0 -5658px}.f-ls[data-v-569f794f],.f-lso[data-v-569f794f]{background-position:0 -5699px}.f-lt[data-v-569f794f],.f-ltu[data-v-569f794f]{background-position:0 -5740px}.f-lu[data-v-569f794f],.f-lux[data-v-569f794f]{background-position:0 -5781px}.f-lv[data-v-569f794f],.f-lva[data-v-569f794f]{background-position:0 -5822px}.f-lby[data-v-569f794f],.f-ly[data-v-569f794f]{background-position:0 -5863px}.f-ma[data-v-569f794f],.f-mar[data-v-569f794f]{background-position:0 -5904px}.f-mc[data-v-569f794f],.f-mco[data-v-569f794f]{background-position:0 -5945px}.f-md[data-v-569f794f],.f-mda[data-v-569f794f]{background-position:0 -5986px}.f-me[data-v-569f794f],.f-mne[data-v-569f794f]{background-position:0 -6027px}.f-maf[data-v-569f794f],.f-mf[data-v-569f794f]{background-position:0 -6068px}.f-mdg[data-v-569f794f],.f-mg[data-v-569f794f]{background-position:0 -6109px}.f-mh[data-v-569f794f],.f-mhl[data-v-569f794f]{background-position:0 -6150px}.f-mk[data-v-569f794f],.f-mkd[data-v-569f794f]{background-position:0 -6191px}.f-ml[data-v-569f794f],.f-mli[data-v-569f794f]{background-position:0 -6232px}.f-mm[data-v-569f794f],.f-mmr[data-v-569f794f]{background-position:0 -6273px}.f-mn[data-v-569f794f],.f-mng[data-v-569f794f]{background-position:0 -6314px}.f-mac[data-v-569f794f],.f-mo[data-v-569f794f]{background-position:0 -6355px}.f-mnp[data-v-569f794f],.f-mp[data-v-569f794f]{background-position:0 -6396px}.f-mq[data-v-569f794f],.f-mtq[data-v-569f794f]{background-position:0 -6437px}.f-mr[data-v-569f794f],.f-mrt[data-v-569f794f]{background-position:0 -6478px}.f-ms[data-v-569f794f],.f-msr[data-v-569f794f]{background-position:0 -6519px}.f-mlt[data-v-569f794f],.f-mt[data-v-569f794f]{background-position:0 -6560px}.f-mu[data-v-569f794f],.f-mus[data-v-569f794f]{background-position:0 -6601px}.f-mdv[data-v-569f794f],.f-mv[data-v-569f794f]{background-position:0 -6642px}.f-mw[data-v-569f794f],.f-mwi[data-v-569f794f]{background-position:0 -6683px}.f-mex[data-v-569f794f],.f-mx[data-v-569f794f]{background-position:0 -6724px}.f-my[data-v-569f794f],.f-mys[data-v-569f794f]{background-position:0 -6765px}.f-moz[data-v-569f794f],.f-mz[data-v-569f794f]{background-position:0 -6806px}.f-na[data-v-569f794f],.f-nam[data-v-569f794f]{background-position:0 -6847px}.f-nc[data-v-569f794f],.f-ncl[data-v-569f794f]{background-position:0 -6888px}.f-ne[data-v-569f794f],.f-ner[data-v-569f794f]{background-position:0 -6929px}.f-nf[data-v-569f794f],.f-nfk[data-v-569f794f]{background-position:0 -6970px}.f-ng[data-v-569f794f],.f-nga[data-v-569f794f]{background-position:0 -7011px}.f-ni[data-v-569f794f],.f-nic[data-v-569f794f]{background-position:0 -7052px}.f-nl[data-v-569f794f],.f-nld[data-v-569f794f]{background-position:0 -7093px}.f-no[data-v-569f794f],.f-nor[data-v-569f794f]{background-position:0 -7134px}.f-np[data-v-569f794f],.f-npl[data-v-569f794f]{background-position:0 -7175px}.f-nr[data-v-569f794f],.f-nru[data-v-569f794f]{background-position:0 -7216px}.f-niu[data-v-569f794f],.f-nu[data-v-569f794f]{background-position:0 -7257px}.f-nz[data-v-569f794f],.f-nzl[data-v-569f794f]{background-position:0 -7298px}.f-om[data-v-569f794f],.f-omn[data-v-569f794f]{background-position:0 -7339px}.f-pa[data-v-569f794f],.f-pan[data-v-569f794f]{background-position:0 -7380px}.f-pe[data-v-569f794f],.f-per[data-v-569f794f]{background-position:0 -7421px}.f-pf[data-v-569f794f],.f-pyf[data-v-569f794f]{background-position:0 -7462px}.f-pg[data-v-569f794f],.f-png[data-v-569f794f]{background-position:0 -7503px}.f-ph[data-v-569f794f],.f-phl[data-v-569f794f]{background-position:0 -7544px}.f-pak[data-v-569f794f],.f-pk[data-v-569f794f]{background-position:0 -7585px}.f-pl[data-v-569f794f],.f-pol[data-v-569f794f]{background-position:0 -7626px}.f-pm[data-v-569f794f],.f-spm[data-v-569f794f]{background-position:0 -7667px}.f-pcn[data-v-569f794f],.f-pn[data-v-569f794f]{background-position:0 -7708px}.f-pr[data-v-569f794f],.f-pri[data-v-569f794f]{background-position:0 -7749px}.f-ps[data-v-569f794f],.f-pse[data-v-569f794f]{background-position:0 -7790px}.f-prt[data-v-569f794f],.f-pt[data-v-569f794f]{background-position:0 -7831px}.f-plw[data-v-569f794f],.f-pw[data-v-569f794f]{background-position:0 -7872px}.f-pry[data-v-569f794f],.f-py[data-v-569f794f]{background-position:0 -7913px}.f-qa[data-v-569f794f],.f-qat[data-v-569f794f]{background-position:0 -7954px}.f-re[data-v-569f794f],.f-reu[data-v-569f794f]{background-position:0 -7995px}.f-ro[data-v-569f794f],.f-rou[data-v-569f794f]{background-position:0 -8036px}.f-rs[data-v-569f794f],.f-srb[data-v-569f794f]{background-position:0 -8077px}.f-ru[data-v-569f794f],.f-rus[data-v-569f794f]{background-position:0 -8118px}.f-rw[data-v-569f794f],.f-rwa[data-v-569f794f]{background-position:0 -8159px}.f-sa[data-v-569f794f],.f-sau[data-v-569f794f]{background-position:0 -8200px}.f-sb[data-v-569f794f],.f-slb[data-v-569f794f]{background-position:0 -8241px}.f-sc[data-v-569f794f],.f-syc[data-v-569f794f]{background-position:0 -8282px}.f-sd[data-v-569f794f],.f-sdn[data-v-569f794f]{background-position:0 -8323px}.f-se[data-v-569f794f],.f-swe[data-v-569f794f]{background-position:0 -8364px}.f-sg[data-v-569f794f],.f-sgp[data-v-569f794f]{background-position:0 -8405px}.f-sh[data-v-569f794f],.f-shn[data-v-569f794f]{background-position:0 -8446px}.f-si[data-v-569f794f],.f-svn[data-v-569f794f]{background-position:0 -8487px}.f-sj[data-v-569f794f],.f-sjm[data-v-569f794f]{background-position:0 -8528px}.f-sk[data-v-569f794f],.f-svk[data-v-569f794f]{background-position:0 -8569px}.f-sl[data-v-569f794f],.f-sle[data-v-569f794f]{background-position:0 -8610px}.f-sm[data-v-569f794f],.f-smr[data-v-569f794f]{background-position:0 -8651px}.f-sen[data-v-569f794f],.f-sn[data-v-569f794f]{background-position:0 -8692px}.f-so[data-v-569f794f],.f-som[data-v-569f794f]{background-position:0 -8733px}.f-sr[data-v-569f794f],.f-sur[data-v-569f794f]{background-position:0 -8774px}.f-ss[data-v-569f794f],.f-ssd[data-v-569f794f]{background-position:0 -8815px}.f-st[data-v-569f794f],.f-stp[data-v-569f794f]{background-position:0 -8856px}.f-slv[data-v-569f794f],.f-sv[data-v-569f794f]{background-position:0 -8897px}.f-sx[data-v-569f794f],.f-sxm[data-v-569f794f]{background-position:0 -8938px}.f-sy[data-v-569f794f],.f-syr[data-v-569f794f]{background-position:0 -8979px}.f-swz[data-v-569f794f],.f-sz[data-v-569f794f]{background-position:0 -9020px}.f-tc[data-v-569f794f],.f-tca[data-v-569f794f]{background-position:0 -9061px}.f-tcd[data-v-569f794f],.f-td[data-v-569f794f]{background-position:0 -9102px}.f-atf[data-v-569f794f],.f-tf[data-v-569f794f]{background-position:0 -9143px}.f-tg[data-v-569f794f],.f-tgo[data-v-569f794f]{background-position:0 -9184px}.f-th[data-v-569f794f],.f-tha[data-v-569f794f]{background-position:0 -9225px}.f-tj[data-v-569f794f],.f-tjk[data-v-569f794f]{background-position:0 -9266px}.f-tk[data-v-569f794f],.f-tkl[data-v-569f794f]{background-position:0 -9307px}.f-tl[data-v-569f794f],.f-tls[data-v-569f794f]{background-position:0 -9348px}.f-tkm[data-v-569f794f],.f-tm[data-v-569f794f]{background-position:0 -9389px}.f-tn[data-v-569f794f],.f-tun[data-v-569f794f]{background-position:0 -9430px}.f-to[data-v-569f794f],.f-ton[data-v-569f794f]{background-position:0 -9471px}.f-tr[data-v-569f794f],.f-tur[data-v-569f794f]{background-position:0 -9512px}.f-tt[data-v-569f794f],.f-tto[data-v-569f794f]{background-position:0 -9553px}.f-tuv[data-v-569f794f],.f-tv[data-v-569f794f]{background-position:0 -9594px}.f-tw[data-v-569f794f],.f-twn[data-v-569f794f]{background-position:0 -9635px}.f-tz[data-v-569f794f],.f-tza[data-v-569f794f]{background-position:0 -9676px}.f-ua[data-v-569f794f],.f-ukr[data-v-569f794f]{background-position:0 -9717px}.f-ug[data-v-569f794f],.f-uga[data-v-569f794f]{background-position:0 -9758px}.f-um[data-v-569f794f],.f-umi[data-v-569f794f]{background-position:0 -9799px}.f-un[data-v-569f794f]{background-position:0 -9840px}.f-us[data-v-569f794f],.f-usa[data-v-569f794f]{background-position:0 -9881px}.f-ury[data-v-569f794f],.f-uy[data-v-569f794f]{background-position:0 -9922px}.f-uz[data-v-569f794f],.f-uzb[data-v-569f794f]{background-position:0 -9963px}.f-va[data-v-569f794f],.f-vat[data-v-569f794f]{background-position:0 -10004px}.f-vc[data-v-569f794f],.f-vct[data-v-569f794f]{background-position:0 -10045px}.f-ve[data-v-569f794f],.f-ven[data-v-569f794f]{background-position:0 -10086px}.f-vg[data-v-569f794f],.f-vgb[data-v-569f794f]{background-position:0 -10127px}.f-vi[data-v-569f794f],.f-vir[data-v-569f794f]{background-position:0 -10168px}.f-vn[data-v-569f794f],.f-vnm[data-v-569f794f]{background-position:0 -10209px}.f-vu[data-v-569f794f],.f-vut[data-v-569f794f]{background-position:0 -10250px}.f-wf[data-v-569f794f],.f-wlf[data-v-569f794f]{background-position:0 -10291px}.f-ws[data-v-569f794f],.f-wsm[data-v-569f794f]{background-position:0 -10332px}.f-xk[data-v-569f794f]{background-position:0 -10373px}.f-ye[data-v-569f794f],.f-yem[data-v-569f794f]{background-position:0 -10414px}.f-myt[data-v-569f794f],.f-yt[data-v-569f794f]{background-position:0 -10455px}.f-za[data-v-569f794f],.f-zaf[data-v-569f794f]{background-position:0 -10496px}.f-zm[data-v-569f794f],.f-zmb[data-v-569f794f]{background-position:0 -10537px}.f-zw[data-v-569f794f],.f-zwe[data-v-569f794f]{background-position:0 -10578px}.f-su[data-v-569f794f],.f-sun[data-v-569f794f]{background-position:0 -10619px}",
    map: undefined,
    media: undefined
  });
};
/* scoped */


var __vue_scope_id__ = "data-v-569f794f";
/* module identifier */

var __vue_module_identifier__ = "data-v-569f794f";
/* functional template */

var __vue_is_functional_template__ = false;
/* style inject shadow dom */

var __vue_component__ = /*#__PURE__*/normalizeComponent({
  render: __vue_render__,
  staticRenderFns: __vue_staticRenderFns__
}, __vue_inject_styles__, __vue_script__, __vue_scope_id__, __vue_is_functional_template__, __vue_module_identifier__, false, undefined, createInjectorSSR, undefined);// Import vue component
// IIFE injects install function into component, allowing component
// to be registered via Vue.use() as well as Vue.component(),

var component = /*#__PURE__*/(function () {
  // Get component instance
  var installable = __vue_component__; // Attach install function executed by Vue.use()

  installable.install = function (Vue) {
    Vue.component('CountryFlag', installable);
  };

  return installable;
})(); // It's possible to expose named exports when writing components that can
// also be used as directives, etc. - eg. import { RollupDemoDirective } from 'rollup-demo';
// export const RollupDemoDirective = directive;
var namedExports=/*#__PURE__*/Object.freeze({__proto__:null,'default': component});// only expose one global var, with named exports exposed as properties of
// that global var (eg. plugin.namedExport)

Object.entries(namedExports).forEach(function (_ref) {
  var _ref2 = _slicedToArray(_ref, 2),
      exportName = _ref2[0],
      exported = _ref2[1];

  if (exportName !== 'default') component[exportName] = exported;
});module.exports=component;
